openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -130,45 +236,147 @@ def __contains__(self, item):
         # Fall back to traditional access
         return list.__contains__(self, item)
 
+    def to_list(self, return_cosmetic_attributes=False):
+        """
+        Render this ParameterList to a normal list, serializing each ParameterType-derived object in it to dict.
+
+        Parameters
+        ----------
+
+        return_cosmetic_attributes : bool, optional. default = False
+            Whether to return non-spec attributes of each ParameterType-derived object.
+
+        Returns
+        -------
+        parameter_list : List[dict]
+            A serialized representation of a ParameterList, with each ParameterType it contains converted to dict.
+        """
+        parameter_list = list()
+
+        for parameter in self:
+            parameter_dict = parameter.to_dict(return_cosmetic_attributes=return_cosmetic_attributes)
+            parameter_list.append(parameter_dict)
+
+        return parameter_list
+
+
 
 # TODO: Rename to better reflect role as parameter base class?
 class ParameterType(object):
     """
     Base class for SMIRNOFF parameter types.
 
     """
-
-    _VALENCE_TYPE = None  # ChemicalEnvironment valence type for checking SMIRKS is conformant
-
-    # TODO: Allow preferred units for each parameter type to be specified and remembered as well for when we are writing out
+    # This list will be used for validating input and detecting cosmetic attributes.
+    _VALENCE_TYPE = None  # ChemicalEnvironment valence type string expected by SMARTS string for this Handler
+    _SMIRNOFF_ATTRIBS = ['smirks'] # Attributes expected per the SMIRNOFF spec.
```

Comment:
This is definitely something that we can do in a separate PR (if we do it at all), but I found tricky to handle about the kwargs defined in class constants in my experience is inheritance. To make a trivial example, you need to re-add `smirks` to the list of `_SMIRNOFF_ATTRIBS` when you inherit from `ParameterType`, but when 2 levels of inheritance and attributes less general than `smirks` are involved, it's very easy to make a mistake that ends in a silent error.

A possible way to take advantage of the Python inheritance implementation to take care of everything could be to use properties instead. In this case, a parameter type could be defined with something like
```python
class BondType(ParameterType):
    length = SmirnoffAttribute(required_unit=unit.angstrom)
    k = SmirnoffAttribute(required_unit=unit.kilocalorie_per_mole / unit.angstrom**2, indexed=True)
    made_up_attr = OptionalAttribute()
```
With this implementation, `ParameterType` can discover required and optional attributes with something like
```python
class ParameterType:
    @classmethod
    def _get_smirnoff_attributes(cls):
        return {name: descriptor for name, descriptor in inspect.getmembers(cls)
                if isinstance(descriptor, SmirnoffAttribute)}
```
I think a design that goes in this direction could make both `ParameterType` and `ParameterHandler` easier to extend.

Reply:
Brilliant. Totally agree. Thanks for the careful writeup. I've turned this into Issue #207 .